pkg/commands/start_test.go

```diff
@@ -16,67 +16,84 @@ package commands
+	t.Logf("Checking for UDP port is running on %v", port)
+	udpListener, err := net.Listen("udp", fmt.Sprintf(":%s", port))
 	if udpListener != nil {
 		defer udpListener.Close()
 	}
-	require.Errorf(t, err, "host-resolver is not listening on UDP port %s", udpPort)
+	require.Errorf(t, err, "host-resolver is not listening on UDP port %s", port)
 
 	output := netstat(t)
 	require.Contains(t, string(output), fmt.Sprintf("%v/host-resolver", cmd.Process.Pid), "Expected the same Pid")
 }
 
-func TestQueryTCP(t *testing.T) {
-	t.Logf("Checking for TCP port on %s", tcpPort)
-	cmd := run(t, "127.0.0.1", tcpPort, udpPort, "host.rd.test=111.111.111.111,host2.rd.test=222.222.222.222")
+func TestQueryStaticHosts(t *testing.T) {
+	port := randomPort()
+	cmd := runHostResovler(t, []string{"-a", "127.0.0.1", "-t", port, "-u", port, "-c", "host.rd.test=111.111.111.111,host2.rd.test=222.222.222.222"})
 	defer cmd.Process.Kill()
 
-	addrs, err := dnsLookup(t, tcpPort, "tcp", "host.rd.test")
+	t.Logf("Checking for TCP port on %s", port)
+	addrs, err := dnsLookup(t, port, "tcp", "host.rd.test")
 	require.NoError(t, err, "Lookup IP failed")
 
 	expected := []net.IP{net.IPv4(111, 111, 111, 111)}
 	require.ElementsMatch(t, ipToString(addrs), ipToString(expected))
+
+	t.Logf("Checking for UDP port on %s", port)
+	addrs, err = dnsLookup(t, port, "udp", "host2.rd.test")
+	require.NoError(t, err, "Lookup IP failed")
+
+	expected = []net.IP{net.IPv4(222, 222, 222, 222)}
+	require.ElementsMatch(t, ipToString(addrs), ipToString(expected))
 }
 
-func TestQueryUDP(t *testing.T) {
-	t.Logf("Checking for UDP port on %s", udpPort)
-	cmd := run(t, "127.0.0.1", tcpPort, udpPort, "host.rd.test=111.111.111.111,host2.rd.test=222.222.222.222")
+func TestQueryUpstreamServer(t *testing.T) {
+	port := randomPort()
+	cmd := runHostResovler(t, []string{"-a", "127.0.0.1", "-t", port, "-u", port, "-s", "[8.8.8.8]"})
 	defer cmd.Process.Kill()
 
-	addrs, err := dnsLookup(t, udpPort, "udp", "host2.rd.test")
+	t.Logf("Resolving via upstream server on [TCP] --> %s", port)
+	addrs, err := dnsLookup(t, port, "tcp", "google.ca")
 	require.NoError(t, err, "Lookup IP failed")
+	require.True(t, len(addrs) > 0, true, "Expect at least an address")
 
-	expected := []net.IP{net.IPv4(222, 222, 222, 222)}
-	require.ElementsMatch(t, ipToString(addrs), ipToString(expected))
+	t.Logf("Resolving via upstream server on [UDP] --> %s", port)
+	addrs, err = dnsLookup(t, port, "udp", "google.ca")
+	require.NoError(t, err, "Lookup IP failed")
+	require.True(t, len(addrs) > 0, true, "Expect at least an address")
 }
 
-func run(t *testing.T, ip, tcpPort, udpPort, hosts string) *exec.Cmd {
-	cmd := exec.Command("/app/host-resolver", "run", "-a", ip, "-t", tcpPort, "-u", udpPort, "-c", hosts)
+func runHostResovler(t *testing.T, args []string) *exec.Cmd {
+	// add run command to the tip
+	args = append([]string{"run"}, args...)
+	// add background process to the tail
+	args = append(args, "&")
```

Comment:
What background process? You're not running the command with `/bin/sh`, there's nothing here to interpret that.

Also, you're using `cmd.Start()` (instead of `cmd.Run()`), it's never blocking anything in the first place…

Reply:
Yeah, not needed at all, it's an artifact of previous attempts.